data-access/seedwork/domain-seedwork/aggregate-root.ts

```diff
@@ -1,46 +1,53 @@
+import { SyncDomainEventPayloadBaseType, SyncDomainEventType } from '../event-bus-seedwork-node/sync-domain-event-bus';
 import { Visa } from '../passport-seedwork/visa';
 import { BaseDomainExecutionContext } from './base-domain-execution-context';
 import { DomainEntity, DomainEntityProps } from './domain-entity';
 import { CustomDomainEvent, DomainEvent } from './domain-event';
 
-export interface RootEventRegistry {
+export interface RootEventRegistry <ContextType extends BaseDomainExecutionContext>{
   addDomainEvent<EventProps, T extends CustomDomainEvent<EventProps>>(event: new (aggregateId: string) => T, props: T['payload']);
   addIntegrationEvent<EventProps, T extends CustomDomainEvent<EventProps>>(event: new (aggregateId: string) => T, props: T['payload']);
+  get context(): ContextType;
   get visa(): Visa;
-  get syncDomainEventClass(): any;
+  addSyncDomainEvent<EventPayloadType extends SyncDomainEventPayloadBaseType, T extends SyncDomainEventType<EventPayloadType>>(
+    event: new () => T, 
+    payload: EventPayloadType
+  ) : void;
 }
 
-export  class AggregateRoot<
-  PropType extends DomainEntityProps,
-  ContextType extends BaseDomainExecutionContext,
-  VisaType extends Visa
-> extends DomainEntity<PropType> implements RootEventRegistry {
+export interface RootEventRegistryForRepo {
+  processSyncDomainEvents(): void;
+}
+
+export  class AggregateRoot<PropType extends DomainEntityProps, ContextType extends BaseDomainExecutionContext, VisaType extends Visa> 
```

Comment:
Reasons for not accepting your suggestions:
1. processSyncDomainEvents method should only be exposed to infrastructure service. This is executed within onSave method inside repo class, and must not be executed by any domain method.
2. additional complexity in processSyncDomainEvents method is introduced to prevent the app from going into an infinite loop, because a sync domain event handler can generate a new series of sync domain events 
3. done